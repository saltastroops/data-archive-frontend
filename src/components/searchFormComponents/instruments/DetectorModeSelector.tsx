import * as React from "react";
import MultiSelectField, {
  AllOption,
} from "../../basicComponents/MultiSelectField";

// DETECTOR MODE
export const COMMON_DETECTOR_MODES = [
  "Normal",
  "Frame Transfer",
  "Slot Mode",
  "Drift Scan",
];

export const RSS_DETECTOR_MODES = ["Shuffle"];
export const SALTICAM_DETECTOR_MODES = [];

interface IDetectorModeSelectorProps {
  detectorModes: string[];
  selectedInstruments?: string[];
  selectedTelescopes?: string[];
  onSelect: (value: any) => void;
}

const detectorModeToDisplay = (instruments: string[], telescopes: string[]) => {
  let detectorModes: string[] = [];
  if (
    instruments.length === 0 ||
    (instruments.some((inst) => inst === "All") &&
      telescopes.some((t) => t === "All" || t === "SALT"))
  ) {
    return RSS_DETECTOR_MODES.concat(SALTICAM_DETECTOR_MODES).concat(
      COMMON_DETECTOR_MODES
    );
  }
  if (telescopes.some((t) => t === "1.9 m" || t === "Lesedi")) {
    return [];
  }
  if (instruments.some((inst) => inst === "RSS")) {
    detectorModes = detectorModes.concat(RSS_DETECTOR_MODES);
  }
  if (instruments.some((inst) => inst === "Salticam")) {
    detectorModes = detectorModes.concat(SALTICAM_DETECTOR_MODES);
  }
  if (instruments.some((inst) => inst === "Salticam" || inst === "RSS")) {
    detectorModes = detectorModes.concat(COMMON_DETECTOR_MODES);
  }

  return detectorModes;
};
/**
 * A form for selecting DetectorModeSelector-related search parameters.
 */
const DetectorModeSelector = (props: IDetectorModeSelectorProps) => {
  const {
    onSelect,
    selectedInstruments,
    selectedTelescopes,
    detectorModes,
  } = props;

  // Function for handling change events
  const select = (e: React.SyntheticEvent<HTMLSelectElement>) => {
    const values = e.currentTarget.selectedOptions;
    onSelect({
      detectorModes: Array.from(values).map((t: any) => {
        return t.text;
      }),
    });
  };
  const DETECTOR_MODE_TO_DISPLAY = detectorModeToDisplay(
    selectedInstruments || [],
    selectedTelescopes || []
  );

  return (
    <div>
      <p>Detector mode</p>
      <MultiSelectField
        name="detectorModes"
        onChange={select}
        value={detectorModes || ["All"]}
      >
        <AllOption />
        {DETECTOR_MODE_TO_DISPLAY.sort().map((i) => (
          <option key={i} value={i}>
            {i}
          </option>
        ))}
      </MultiSelectField>
    </div>
  );
};

export default DetectorModeSelector;
