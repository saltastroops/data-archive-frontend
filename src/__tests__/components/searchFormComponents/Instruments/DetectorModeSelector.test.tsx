import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import DetectorModeSelector from "../../../../components/searchFormComponents/instruments/DetectorModeSelector";

const onSelect = jest.fn() as any;
describe("Detector Mode Selector", () => {
  it("should be defined", () => {
    expect(
      mount(<DetectorModeSelector detectorModes={[]} onSelect={onSelect} />)
    ).toBeDefined();
  });

  it("should render correctly", () => {
    // Use mount instead of shallow for better snapshots
    expect(
      toJson(
        mount(<DetectorModeSelector detectorModes={[]} onSelect={onSelect} />)
      )
    ).toMatchSnapshot();
  });
});
