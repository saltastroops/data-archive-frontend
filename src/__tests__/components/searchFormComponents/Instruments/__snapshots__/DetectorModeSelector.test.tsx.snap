// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detector Mode Selector should render correctly 1`] = `
.c0 {
  width: 95%;
}

<DetectorModeSelector
  detectorModes={Array []}
  onSelect={[MockFunction]}
>
  <div>
    <p>
      Detector mode
    </p>
    <MultiSelectField
      name="detectorModes"
      onChange={[Function]}
      value={Array []}
    >
      <styled.div
        className="control select is-multiple "
      >
        <StyledComponent
          className="control select is-multiple "
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": false,
                "lastClassName": "c0",
                "rules": Array [
                  "
  width: 95%;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "div",
              "toString": [Function],
              "usesTheme": false,
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <div
            className="c0 control select is-multiple "
          >
            <select
              multiple={true}
              name="detectorModes"
              onChange={[Function]}
              size={6}
              style={
                Object {
                  "width": "216px",
                }
              }
              value={Array []}
            >
              <AllOption>
                <option
                  value="All"
                >
                  All
                </option>
              </AllOption>
              <option
                key="Drift Scan"
                value="Drift Scan"
              >
                Drift Scan
              </option>
              <option
                key="Frame Transfer"
                value="Frame Transfer"
              >
                Frame Transfer
              </option>
              <option
                key="Normal"
                value="Normal"
              >
                Normal
              </option>
              <option
                key="Shuffle"
                value="Shuffle"
              >
                Shuffle
              </option>
              <option
                key="Slot Mode"
                value="Slot Mode"
              >
                Slot Mode
              </option>
            </select>
          </div>
        </StyledComponent>
      </styled.div>
    </MultiSelectField>
  </div>
</DetectorModeSelector>
`;
